engines/ic3ia.cpp

```diff
@@ -0,0 +1,369 @@
+/*********************                                                  */
+/*! \file ic3ia.cpp
+** \verbatim
+** Top contributors (to current version):
+**   Makai Mann
+** This file is part of the pono project.
+** Copyright (c) 2019 by the authors listed in the file AUTHORS
+** in the top-level source directory) and their institutional affiliations.
+** All rights reserved.  See the file LICENSE in the top-level source
+** directory for licensing information.\endverbatim
+**
+** \brief IC3 via Implicit Predicate Abstraction (IC3IA) implementation
+**        based on
+**
+**        IC3 Modulo Theories via Implicit Predicate Abstraction
+**            -- Alessandro Cimatti, Alberto Griggio,
+**               Sergio Mover, Stefano Tonetta
+**
+**        and the open source implementation:
+**
+**        https://es-static.fbk.eu/people/griggio/ic3ia/index.html
+**/
+
+#include "engines/ic3ia.h"
+
+#include "utils/logger.h"
+#include "utils/term_analysis.h"
+
+using namespace smt;
+using namespace std;
+
+namespace pono {
+
+IC3IA::IC3IA(Property & p, SolverEnum se)
+    : super(p, se), abs_ts_(ts_.solver()), ia_(ts_, abs_ts_)
+{
+  initialize();
+}
```

Comment:
I am not sure if I understand the issue with `initialize`.
we can call the initialize method of the super class: `super::initializer()`. Isn't this serve the purpose?

Reply:
Right, so I just think it's kinda messy. Basically we have this class hierarchy: `Prover -> MBIC3 -> IC3IA`. You can instantiate `MBIC3` on its own, so it needs to have an `initialize` in its constructor. `IC3IA` also has to have its own `initialize` in its constructor. They're similar, but might initialize things slightly differently. One example difference is that the trans label needs to refer to either the concrete or abstract transition relation. However, in the `MBIC3::initialize` method, you can't know if you're instantiating a "raw" `MBIC3` or if it's just a base for an `IC3IA` class. I.e. `virtual` method calls in a base class constructor _don't_ work as expected. We can discuss this more offline.